File: src/main/java/levels/cafe/components/ground/GroundTile.java
package levels.cafe.components.ground;

import engine.graphics.Texture;
import engine.graphics.VertexArray;
import engine.maths.Vector3f;
import engine.AbstractComponent;

import java.util.Objects;

public class GroundTile extends AbstractComponent {

    private final static float Z_LEVEL = 0.0f;
    public final static float VERTICAL_OFFSET = 1;
    public final static float HORIZONTAL_OFFSET = 2;

    private final Vector3f startVector;
    private AbstractFurniture furniture;

    public GroundTile(Vector3f startVector) {
        this.startVector = startVector;
        vertexArray = generateVertexArray();
        texture = generateTexture();
    }

    @Override
    protected VertexArray generateVertexArray() {
        float[] vertices = new float[]{
                startVector.x, startVector.y, Z_LEVEL,
                startVector.x + HORIZONTAL_OFFSET, startVector.y + VERTICAL_OFFSET, Z_LEVEL,
                startVector.x, startVector.y + HORIZONTAL_OFFSET, Z_LEVEL,
                startVector.x - HORIZONTAL_OFFSET, startVector.y + VERTICAL_OFFSET, Z_LEVEL,
        };

        byte[] indices = new byte[]{
                0, 1, 2,
                0, 2, 3
        };

        float[] tcs = new float[]{
                0.5f, 0,
                1, -0.5f,
                0.5f, -1,
                0, -0.5f
        };
        return new VertexArray(vertices, indices, tcs);
    }

    @Override
    protected Texture generateTexture() {
        return new Texture("res/room/ground_tiles/bottom_tile_0003.PNG");
    }


    public AbstractFurniture getFurniture() {
        return furniture;
    }

    public void setFurniture(AbstractFurniture furniture) {
        this.furniture = furniture;
    }

    public Vector3f getStartVector() {
        return startVector;
    }

    public boolean hasFurniture() {
        return !Objects.isNull(furniture);
    }
}

File: src/main/java/levels/cafe/components/ground/GroundTile.kt
package levels.cafe.components.ground

import engine.graphics.Texture
import engine.graphics.TextureLoader
import engine.graphics.VertexArray
import engine.maths.Vector3f
import engine.AbstractComponent

import java.util.Objects

class GroundTile(val startVector: Vector3f) : AbstractComponent() {
    var furniture: AbstractFurniture? = null

    init {
        vertexArray = generateVertexArray()
        texture = generateTexture()
    }

    override fun generateVertexArray(): VertexArray {
        val vertices = floatArrayOf(startVector.x, startVector.y, Z_LEVEL, startVector.x + HORIZONTAL_OFFSET, startVector.y + VERTICAL_OFFSET, Z_LEVEL, startVector.x, startVector.y + HORIZONTAL_OFFSET, Z_LEVEL, startVector.x - HORIZONTAL_OFFSET, startVector.y + VERTICAL_OFFSET, Z_LEVEL)

        val indices = byteArrayOf(0, 1, 2, 0, 2, 3)

        val tcs = floatArrayOf(0.5f, 0f, 1f, -0.5f, 0.5f, -1f, 0f, -0.5f)
        return VertexArray(vertices, indices, tcs)
    }

    override fun generateTexture(): Texture {
        return TextureLoader.getTexture("res/room/ground_tiles/bottom_tile_0003.PNG")
    }

    fun hasFurniture(): Boolean {
        return !Objects.isNull(furniture)
    }

    companion object {
        private val Z_LEVEL = 0.0f
        val VERTICAL_OFFSET = 1f
        val HORIZONTAL_OFFSET = 2f
    }
}
